File: examples/AFTI16_example.jl
using PiMPC
using LinearAlgebra
using Printf
using Plots

# ============================================
#           AFTI-16 Aircraft System
# ============================================

function build_AFTI16(Ts=0.05)
    # Continuous-time system matrices
    As = [-0.0151  -60.5651   0.0      -32.174;
          -0.0001   -1.3411   0.9929    0.0;
           0.00018  43.2541  -0.86939   0.0;
           0.0       0.0      1.0       0.0]

    Bs = [-2.516   -13.136;
          -0.1689   -0.2514;
         -17.251    -1.5766;
           0.0       0.0]

    # Discretization (zero-order hold)
    nx, nu = 4, 2
    M = exp([As Bs; zeros(nu, nx) zeros(nu, nu)] * Ts)
    A = M[1:nx, 1:nx]
    B = M[1:nx, nx+1:end]

    # Output matrix: angle of attack and pitch angle
    C = [0.0 1.0 0.0 0.0;    # Angle of attack (y1)
         0.0 0.0 0.0 1.0]    # Pitch angle (y2)

    return A, B, C
end

# ============================================
#       Closed-Loop MPC Simulation
# ============================================

function simulate_closed_loop(model, A, B, C, x0, u0, yref_traj, uref, w, Nsim)
    nx, nu = size(B)
    ny = size(C, 1)

    # Storage for trajectories
    x_hist = zeros(nx, Nsim + 1)
    y_hist = zeros(ny, Nsim + 1)
    u_hist = zeros(nu, Nsim)
    iter_hist = zeros(Int, Nsim)

    x_hist[:, 1] = x0
    y_hist[:, 1] = C * x0
    x_current = copy(x0)
    u_prev = copy(u0)

    println("Running closed-loop simulation...")

    for k in 1:Nsim
        # Get current reference
        yref = yref_traj[:, k]

        # Solve MPC problem
        results = solve!(model, x_current, u_prev, yref, uref, w; verbose=false)

        # Extract first control input
        u_apply = results.u[:, 1]

        # Apply control to system (simulate one step)
        x_next = A * x_current + B * u_apply
        y_next = C * x_next

        # Store results
        u_hist[:, k] = u_apply
        x_hist[:, k + 1] = x_next
        y_hist[:, k + 1] = y_next
        iter_hist[k] = results.info.iterations

        # Update for next iteration
        x_current = x_next
        u_prev = u_apply
        @info "Step $k"
        @info "u = $(u_apply)"
        @info "x = $(x_next)"
    end

    println("Simulation completed.")
    println("-" ^ 40)
    @printf("  Total steps: %d\n", Nsim)
    @printf("  Average iterations: %.1f\n", sum(iter_hist) / Nsim)

    return x_hist, y_hist, u_hist, iter_hist
end

# ============================================
#           Main Test
# ============================================

println("=" ^ 50)
println("  PiMPC - AFTI-16 Closed-Loop Simulation")
println("=" ^ 50)

# Build system
Ts = 0.05
A, B, C = build_AFTI16(Ts)
nx, nu, ny = 4, 2, 2

# Create and setup MPC model
Np = 5
model = Model()
setup!(model;
    A = A, B = B, C = C, Np = Np,
    Wy = 100.0 * diagm(ones(ny)),
    Wu = 0.0 * diagm(ones(nu)),
    Wdu = 0.1 * diagm(ones(nu)),
    umin = -25.0 * ones(nu),
    umax = 25.0 * ones(nu),
    xmin = [-Inf, -0.5, -Inf, -100.0],
    xmax = [Inf, 0.5, Inf, 100.0],
    rho = 1.0,
    tol = 1e-6,
    maxiter = 10000,
    precond = true,
    accel = true,
    device = :cpu
)

# Initial conditions
x0 = zeros(nx)
u0 = zeros(nu)

# Simulation length
Nsim = 200

# Time-varying reference trajectory
# y1 (angle of attack): always 0
# y2 (pitch angle): 10 deg for first 100 steps, then 0
yref_traj = [zeros(1, Nsim);
             [10.0 * ones(1, 100) zeros(1, 100)]]

uref = zeros(nu)
w = zeros(nx)

# Run simulation
x_hist, y_hist, u_hist, iter_hist = simulate_closed_loop(
    model, A, B, C, x0, u0, yref_traj, uref, w, Nsim
)

# ============================================
#           Plot Results
# ============================================

println("\nGenerating plots...")

t_y = (0:Nsim) * Ts  # Time vector for states/outputs
t_u = (0:Nsim-1) * Ts  # Time vector for inputs

# Top: Trajectory (4x1 subplot)
p1 = plot(t_y, [yref_traj[2, :]; 0], label="Reference",
          ylabel="y2 (deg)", lw=2, ls=:dash, color=:black, legend=:topright)
plot!(p1, t_y, y_hist[2, :], label="Pitch Angle", lw=2, color=:red)

p2 = plot(t_y, y_hist[1, :], ylabel="y1 (deg)", lw=2, color=:red, legend=false)

p3 = plot(t_u, u_hist[1, :], ylabel="u1 (deg)", lw=2, color=:blue, legend=false)

p4 = plot(t_u, u_hist[2, :], ylabel="u2 (deg)", lw=2, color=:blue, legend=false)

# Bottom: Iterations per step
p5 = plot(1:Nsim, iter_hist, xlabel="Step", ylabel="Iterations",
          lw=2, color=:green, legend=false)

# Combine: trajectory on top (4 rows), iterations on bottom (1 row)
fig = plot(p1, p2, p3, p4, p5, layout=@layout([a; b; c; d; e]), size=(800, 800))
display(fig)

println("\nPress Enter to exit...")
readline()

File: examples/afti16_example.py
import numpy as np

from pimpc import Model


def build_afti16(ts=0.05):
    As = np.array(
        [
            [-0.0151, -60.5651, 0.0, -32.174],
            [-0.0001, -1.3411, 0.9929, 0.0],
            [0.00018, 43.2541, -0.86939, 0.0],
            [0.0, 0.0, 1.0, 0.0],
        ]
    )
    Bs = np.array(
        [
            [-2.516, -13.136],
            [-0.1689, -0.2514],
            [-17.251, -1.5766],
            [0.0, 0.0],
        ]
    )

    nx, nu = 4, 2
    try:
        from scipy.linalg import expm
    except ImportError as exc:
        raise RuntimeError("SciPy is required for the AFTI-16 example.") from exc

    M = expm(np.block([[As, Bs], [np.zeros((nu, nx)), np.zeros((nu, nu))]]) * ts)
    A = M[:nx, :nx]
    B = M[:nx, nx:]

    C = np.array(
        [
            [0.0, 1.0, 0.0, 0.0],
            [0.0, 0.0, 0.0, 1.0],
        ]
    )

    return A, B, C


def simulate_closed_loop(model, A, B, C, x0, u0, yref_traj, uref, w, nsim):
    nx, nu = B.shape
    ny = C.shape[0]

    x_hist = np.zeros((nx, nsim + 1))
    y_hist = np.zeros((ny, nsim + 1))
    u_hist = np.zeros((nu, nsim))
    iter_hist = np.zeros(nsim, dtype=int)

    x_hist[:, 0] = x0
    y_hist[:, 0] = C @ x0
    x_current = x0.copy()
    u_prev = u0.copy()

    print("Running closed-loop simulation...")

    for k in range(nsim):
        yref = yref_traj[:, k]
        results = model.solve(x_current, u_prev, yref, uref, w, verbose=False)

        u_apply = results.u[:, 0]
        x_next = A @ x_current + B @ u_apply
        y_next = C @ x_next

        u_hist[:, k] = u_apply
        x_hist[:, k + 1] = x_next
        y_hist[:, k + 1] = y_next
        iter_hist[k] = results.info.iterations

        x_current = x_next
        u_prev = u_apply

    print("Simulation completed.")
    print("-" * 40)
    print(f"  Total steps: {nsim}")
    print(f"  Average iterations: {np.mean(iter_hist):.1f}")

    return x_hist, y_hist, u_hist, iter_hist


def main():
    print("=" * 50)
    print("  PiMPC - AFTI-16 Closed-Loop Simulation")
    print("=" * 50)

    ts = 0.05
    A, B, C = build_afti16(ts)
    nx, nu, ny = 4, 2, 2

    Np = 5
    model = Model()
    model.setup(
        A=A,
        B=B,
        C=C,
        Np=Np,
        Wy=100.0 * np.eye(ny),
        Wu=0.0 * np.eye(nu),
        Wdu=0.1 * np.eye(nu),
        umin=-25.0 * np.ones(nu),
        umax=25.0 * np.ones(nu),
        xmin=np.array([-np.inf, -0.5, -np.inf, -100.0]),
        xmax=np.array([np.inf, 0.5, np.inf, 100.0]),
        rho=1.0,
        tol=1e-6,
        maxiter=10000,
        precond=True,
        accel=True,
        device="cpu",
    )

    x0 = np.zeros(nx)
    u0 = np.zeros(nu)
    nsim = 200

    yref_traj = np.vstack([np.zeros(nsim), np.concatenate([10.0 * np.ones(100), np.zeros(100)])])
    uref = np.zeros(nu)
    w = np.zeros(nx)

    x_hist, y_hist, u_hist, iter_hist = simulate_closed_loop(
        model, A, B, C, x0, u0, yref_traj, uref, w, nsim
    )

    try:
        import matplotlib.pyplot as plt
    except ImportError as exc:
        raise RuntimeError("matplotlib is required to plot results.") from exc

    t_y = np.arange(nsim + 1) * ts
    t_u = np.arange(nsim) * ts

    fig, axes = plt.subplots(5, 1, figsize=(8, 8), sharex=False)

    axes[0].plot(t_y, np.concatenate([yref_traj[1, :], [0.0]]), label="Reference", linestyle="--", color="black")
    axes[0].plot(t_y, y_hist[1, :], label="Pitch Angle", color="red")
    axes[0].set_ylabel("y2 (deg)")
    axes[0].legend(loc="upper right")

    axes[1].plot(t_y, y_hist[0, :], color="red")
    axes[1].set_ylabel("y1 (deg)")

    axes[2].plot(t_u, u_hist[0, :], color="blue")
    axes[2].set_ylabel("u1 (deg)")

    axes[3].plot(t_u, u_hist[1, :], color="blue")
    axes[3].set_ylabel("u2 (deg)")

    axes[4].plot(np.arange(nsim) + 1, iter_hist, color="green")
    axes[4].set_xlabel("Step")
    axes[4].set_ylabel("Iterations")

    plt.tight_layout()
    plt.savefig("example_afti16.png")
    plt.show()


if __name__ == "__main__":
    main()
